src/event_loop.rs

```diff
@@ -278,23 +279,33 @@ impl<T> EventLoop<T> {
         EventLoopBuilder::<T>::with_user_event().build()
     }
 
-    /// Hijacks the calling thread and initializes the winit event loop with the provided
-    /// closure. Since the closure is `'static`, it must be a `move` closure if it needs to
+    /// Runs the event loop in the calling thread and calls the given `event_handler` closure
+    /// to dispatch any window system events.
+    ///
+    /// Since the closure is `'static`, it must be a `move` closure if it needs to
     /// access any data from the calling context.
     ///
     /// See the [`ControlFlow`] docs for information on how changes to `&mut ControlFlow` impact the
     /// event loop's behavior.
     ///
-    /// Any values not passed to this function will *not* be dropped.
-    ///
     /// ## Platform-specific
     ///
     /// - **X11 / Wayland:** The program terminates with exit code 1 if the display server
     ///   disconnects.
+    /// - **iOS:** Will never return to the caller and so values not passed to this function will
+    ///   *not* be dropped before the process exits.
+    /// - **Web:** Will _act_ as if it never returns to the caller by throwing a Javascript exception
+    ///   (that Rust doesn't see) that will also mean that any values not passed to this function
+    ///   will *not* be dropped.
```

Comment:
@rib I think that describes what you thought is why the distinction is important.
Unfortunately, because Rust doesn't have Wasm unwinding support, calling `panic` wouldn't drop any values not passed here as well.
As I said below, no difference.

Though maybe it's important to tell the user that even though they might see a JS exception, it won't call the Rust panic handler.

Reply:
```suggestion
    /// - **Web:** Will _act_ as if it never returns to the caller by throwing a JavaScript exception,
    ///   which will **not** trigger the Rust panic handler. As the Wasm target on Rust doesn't
    ///   have unwinding support, any values not passed to this function will **not** be dropped.
```

Reply:
I'm not sure about saying that the reason nothing is dropped is due to Rust not having wasm target unwinding support.

This trick is very dependent on Rust not knowing that the machine it runs on has it's own form of exception.

If the Wasm target on Rust did have some kind of unwinding support then we would presumably actually need to disable that here or else it wouldn't be possible to use this trick anymore.

If Rust could somehow catch unwinding done by Wasm then it still wouldn't be allowed to drop things from the scope outside the Winit event loop (i.e. couldn't drop things from the `main()` function) because that would break Rust's lifetime rules.

The reason those things aren't dropped is because a trick is being use to avoid executing the remainder of the `main()` function to provide the illusion that the function never returns but actually it basically returns immediately but done in a way that Rust can't observe.

Conceptually you end up with an `EventLoop::run()` function that runs forever and so it's similar to the iOS backend where an `NSApplication::run()` never returns and that means the second half of the main function doesn't run and so the code for Dropping things left in the main() function doesn't run.

Reply:
> Maybe we should disable `run` for these platforms instead? Users will have glue code for them regardless iirc.

I don't think it's bad, it's not undefined behavior, just memory leaks. I think if we want to maintain our cross-platform goals we should leave it.

---

@rib I'm not sure if we are disagreeing on something here, but I would like to clarify what I think here:
If Rust supported the Wasm exception handling proposal, the thrown exception could be caught right back into Rust which would do it's thing: clean up the calling function.

Currently we are just abusing the fact that Rust doesn't emit any `catch` instruction and does nothing when a JS exception is thrown and therefor the function is not continued. It's probably more accurate to say that because browsers don't detect the calling Rust function trying to (literally) `catch` the exception, they just continue past it and that's why the Rust function is never resumed (or unwound through by Rust after catching it).

> If Rust could somehow catch unwinding done by Wasm then it still wouldn't be allowed to drop things from the scope outside the Winit event loop (i.e. couldn't drop things from the `main()` function) because that would break Rust's lifetime rules.

It's what it would have to do though. Just out of curiosity, why would that break Rust's lifetime rules? AFAIK things defined in the `fn main()` function in Rust don't get `'static` lifetimes.

I understand where you are coming from, you are just saying how it is.
But I'm talking about how it should be, that this isn't implemented and that if it would this wouldn't be happening.

Personally I don't like to use the words "Rust doesn't see" or "Rust can't observe", this isn't really a technical description/explanation. So I am opting to describe what's lacking instead, because the status-quo is much harder to describe accurately.

But I can try:
> As the JavaScript environment won't continue back execution into Rust, any values not passed to this function will **not** be dropped.

> As Rust can't observe the exception and resume execution or unwind, any values not passed to this function will **not** be dropped.

> As the calling function is never resumed, any values not passed to this function will **not** be dropped.

I think I can live with the last one. WDYT?

Reply:
I just realized in #2900 that what would be far more important to document here is that it never returns and therefor doesn't execute any code afterwards.
Currently we are just talking about "dropping", which isn't nearly as important.

As I pointed out in #2900, I believe that this would be a bad change for Winit because `run()` would be fundamentally not cross-platform anymore, considering this isn't really compatible with Web.

Reply:
> I believe that this would be a bad change for Winit because run() would be fundamentally not cross-platform anymore, considering this isn't really compatible with Web.

As also noted in #2900 follow up - this change is strictly more portable that the previous `run() -> !` API since this lets us fix https://github.com/rust-windowing/winit/issues/2709

It's not perfect for Web and iOS but it is portable + usable on Web and iOS.

We can either provide a `run_noreturn` API that could provide better ergonomics for Web + iOS - or we could revert to using extensions on all platforms.

I think it would be unfortunate to require an extension for all platforms since it will end up making it slightly more awkward  to write portable code (considering `run() -> Result` is actually is portable separate from the caveat that it doesn't return on Web + iOS, it's valid for a function to logically run indefinitely).

Reply:
I think it's actually fine because rust doesn't guarantee the destructoros to run, hence it's not a UB in the first place if you look at it.

Reply:
> I think it's actually fine because rust doesn't guarantee the destructoros to run, hence it's not a UB in the first place if you look at it.

Not running destructors isn't an issue, I agree. This is why `run() -> !` was fine. But the way `run() -> Result` is supposed to work is that it continues the function after finishing, which is not the case on Web. So the API is imho a big footgun. Especially because it doesn't work the same across platforms. E.g. you could have some cleanup going on afterwards but you don't notice that the code doesn't get executed on Web.

> I think it would be unfortunate to require an extension for all platforms since it will end up making it slightly more awkward to write portable code (considering `run() -> Result` is actually is portable separate from the caveat that it doesn't return on Web + iOS, it's valid for a function to logically run indefinitely).

I agree. I would like to suggest to not make it an extension method, but just make the APIs only available on certain platforms, ergo put the methods behind `cfg` guards instead of extension traits.

- `run() -> Result` available on all platforms except Web and iOS.
- `run_noreturn() -> !` available on all platforms except Android.
- `spawn()` available on Web only (that one could stay behind the extension trait).

Does that sound right?

Reply:
`run() -> Result` should still be made available on Web and iOS imho - even if we add a preferable `run_noreturn() -> !` for Web / iOS

In terms of practically writing portable code then the option of having `run() -> Result` on _all_ platforms is nice to have.

In 99.999% of cases the only code after `run()` is just going to return a status from `main()` or log an error for a failure.

Theoretically someone could write code code after `run()` that tries to save some state, which is in itself something that's awkward to do in a portable way anyway, and then that same developer could go so far as writing an entrypoint for web and _still_ not have considered the possibility that the save state code wont run, and the unlikely consequence is that they realize their mistake the first time they test their app in the browser.

This kind of issue is something that exists for any language that doesn't happen to support the notion of `-> !`

Their is no guarantee that a function that specifies a return type needs to actually return in a finite amount of time and so it's always valid for that function to just never return, and it can be documented to never return on Web an iOS.

I don't see any big foot gun here - it's a spud gun at best.

Reply:
> In terms of practically writing portable code then the option of having `run() -> Result` on _all_ platforms is nice to have.

I very much disagree with the implication here. Yes, it **would** be nice. But this function is not portable, it does very different things on these platforms, that's why I call it a footgun.

I don't disagree with your arguments that most of the time this isn't an issue and it's not particularly hard to debug. But I do disagree that this is how we should design our API in general.
I would argue that Rust is not that type of language, if we return a `Result` then we should return a `Result` and not a `!`. The argument that most languages don't have a `!` type doesn't make sense to me, this is a Rust library, Rust has a `!` type.

An alternative suggestion is that `run() -> Result` could just return an error on Web. That would make sense to me and would immediately notify a developer that this API doesn't work on Web.

I understand that there is a high motivation to make a cross-platform API that works on **every** platform and I very much share that sentiment (me sharing that sentiment or not isn't actually relevant, Winit is built on it), but exposing an API that is misleading or at least doesn't reflect the high-quality of cross-platform parity we try and put into the rest of the API is not something I personally can agree with.

Reply:
> An alternative suggestion is that run() -> Result could just return an error on Web. That would make sense to me and would immediately notify a developer that this API doesn't work on Web.

that would be worse I think because then you will only find out a runtime that your code isn't going to do anything - at least if the web backend doesn't want to expose `run() -> Result` then we may as well  just not make the API available at all at compile time.

> I understand that there is a high motivation to make a cross-platform API that works on every platform and I very much share that sentiment 

I think it will be less ergonomic for application developers wanting to write portable code to not have at least one `run()` API that they can assume is available across _all_ platforms - and `run() -> Result` can provide that.

> but exposing an API that is misleading or at least doesn't reflect the high-quality of cross-platform parity we try and put into the rest of the API is not something I personally can agree with.

The API isn't misleading though, it's documented with a platform-specific caveat for Web and iOS like lots of Winit APIs, and  it behaves completely consistently with all other platforms except for the fact that the function will never return. Functions are allowed to never return, they don't _have_ to use `->!` to exhibit that behaviour.

Considering the amount of effort I have put into the discussions and work for this branch which I see as a pretty decent step forwards for trying to improve the consistency of winit across platforms, it's honestly pretty annoying to have you dismissing my pov because it doesn't meet some subjective bar of 'quality' that we are striving for in Winit.

I guess this is a classic bike shed issue and I suppose I've repeatedly shared how I see this, and personally I don't see this in the same way as you. I personally think my view is pragmatic, but that's subjective. I understand that this is a divisive detail, so I'm going to step back.

I'm clearly quite opinionated that having a portable `run() -> Result` would be my preference but it's not going to be the end of the world if we no longer have a portable `run()` API.

Reply:
I don't want to annoy rib anymore with this topic, but I'm not gonna continue pushing on this alone either.
Ofc we can figure this issue out in a follow-up PR.

I do believe that a `run() -> Result` function that behaves quite differently depending on the platform is not an acceptable API surface.
So I'm gonna try to elaborate where I'm coming from here.

When I think about designing APIs I'm trying to think about how the user looks at it:
1. The name of the function including it's path.
2. The function signature, parameter and return types, generics, etc.
3. The documentation and context around it.

If a function is called `as_str()` but returns a `i32`, that probably wouldn't make sense.
If a function `to_string() -> String` crates a window, it's name is probably misleading, even if that's documented.
These are pretty trivial examples, just trying to make a point here about the order of things.

We are calling a function `run() -> Result`, which on some platforms returns and on some others it doesn't.
I find the difference between these two behaviors quite big to simply use the same function for it, especially if it's name and signature has no indication about any difference.

I would very strongly prefer a portable `run()` method as well, but I don't want it at all costs. where I set this bar and what I perceive as "cost" is obviously subjective.

To summarize: my main problem is that documentation isn't the only indicator we should rely on to explain function behavior. I would like to tell users somehow what they can expect, the `run()` function doesn't push users to read every detail in the documentation about it, it's non-obvious.

So I was thinking on how to address this problem:
1. Name it something like `run_cross()`, the word "cross" here is non-obvious and would push users towards reading the documentation around it.
2. Hide it behind a cross-platform extension trait, which would hopefully also prompt users to check why this is "hidden" behind a trait.
3. At a minimum, I think it should be very clear from the first sentence that this method doesn't behave the same between platforms.

As I said at the start, I'm not gonna push this any further if I'm the only one having this opinion.

Reply:
> I don't want to annoy rib anymore with this topic, but I'm not gonna continue pushing on this alone either.

It probably shouldn't annoy me as much as it does.

I think it's just one of those things where there's no "right" answer - which in itself is fine. At this point I think I'm less concerned with the actual design decision but just find it a bit annoying ending up having the same discussions repeatedly from scratch - which I also appreciate isn't your fault. I implemented this change based on a different conclusion than the one your advocating for now.

I also strongly believe in considering the application developer experience. It's all about trade offs here and I don't have the same view as you regarding the perception of a function can run indefinitely - I find that completely acceptable personally. I understand the fear around developers potentially trying to do things after the function returns but I've only ever really seen that as an academic concern personally and I don't tend to like being overly focused on things when there's no evidence for a real problem. The academic nature of this concern is also what contributes to there being no right answer. It's easy to either strongly assert that you believe lots of developers will either fall afoul of this issue or assert the opposite.

It's probably best for @kchibisov to just make a decision.

> Name it something like run_cross(), the word "cross" here is non-obvious and would push users towards reading the documentation around it.

Could call it `run_maybe_return()` It's a bit over-the-top but couldn't really be much clearer that the function might not return. 

Overall though I probably also overstated my preference for `run() -> Result` being available for all platforms - conflating with my annoyance at repeating this discussion.

In practice I think Web + iOS + Android apps are likely to have bespoke entrypoints so if separate `run` APIs are exposed for certain platforms that's probably not going to be that difficult for developers to handle.

Reply:
> I understand the fear around developers potentially trying to do things after the function returns but I've only ever really seen that as an academic concern personally and I don't tend to like being overly focused on things when there's no evidence for a real problem.

Just want to point out that we did remove `run_return()` and replaced it with `run() -> Result`.
At least to me this isn't an academic concern because I have used `run_return()` as well to do cleanup afterwards (I know, one-person-statistics are great).

> In practice I think Web + iOS + Android apps are likely to have bespoke entrypoints so if separate `run` APIs are exposed for certain platforms that's probably not going to be that difficult for developers to handle.

I think this would be an argument for this only being an academic concern. But this isn't the case for Web, `fn main()` support in Web is known and widely used, even we use it ourselves in the examples.

> Could call it `run_maybe_return()` It's a bit over-the-top but couldn't really be much clearer that the function might not return.

I would be perfectly happy with that.

Reply:
> Just want to point out that we did remove run_return() and replaced it with run() -> Result.

It's really more just a convenience that the `run` API change is being done in the same PR as the `pump_events` change since we generally discussed the `run` API changes at the same time as agreeing the `pump_events` changes.

I'd say that `run_return` is being replaced by `run_ondemand` and `pump_events`.

The change to `run` is essentially orthogonal, but yeah it does also address another use case for `run_return` which was to get an exit status.

> fn main() support in Web is known and widely used, even we use it ourselves in the examples.

right, but the examples that rely on a regular `fn main()` to run on web are probably also a counter example for requiring a special case `run` functions for certain platforms because it's convenient that currently is possible to write a `main()` function using `run` across 

 `run` API for web, because that code is more likely to have not been written with web specifically in mind.


but the 
right, it's convenient. I see the examples as a counter example for wanting to have a sp

Reply:
> Just want to point out that we did remove run_return() and replaced it with run() -> Result.

It's really more just a convenience that the `run` API change is being done in the same PR as the `pump_events` change since we generally discussed the `run` API changes at the same time as agreeing the `pump_events` changes.

I'd say that `run_return` is being replaced by `run_ondemand` and `pump_events`.

The change to `run` is essentially orthogonal, but yeah it does also address another use case for `run_return` which was to get an exit status.

> fn main() support in Web is known and widely used, even we use it ourselves in the examples.

right, but the examples that rely on a regular `fn main()` to run on web are probably also a counter example for requiring a special case `run` functions for certain platforms because it's convenient to be able to write a `main()` function using a common `run` function across all platforms.

Reply:
In terms of what I think is "academic" I should clarify that I'm not saying that writing clean up code is academic on platforms that can return from `run()`

The point at which I think it becomes academic though is with worrying that developers will write important code that they think will run after the run loop returns and they will then end up with a serious problem when their code runs on web or iOS because these platforms don't return.

It's really hard for me to imagine a realistic situation where there's a real problem here. 

The first hurdle for this being a problem is coming up with some cleanup/pos-run code that would also need to be meaningfully portable to web - because I'd expect that would be the first hint that you need to think about special casing web when you realize your app's random log file flushing or whatever doesn't make sense on web anyway.

Even if you really did get as far as writing some portable clean up code I would expect that it still wouldn't logically make sense to run on web in most cases anyway - the concept of having cleanup code for a web page sounds very unusual to me, I can't really imagine developers having any expectation of running clean up code after `run` returns on web.

The most likely code I can think of, that would also be portable to web / iOS, would be benign and it wouldn't matter that it's not reached.

Reply:
I'm not really seeing it that way, as I said, I myself run cleanup code, which happens to also be portable, which isn't really a surprise because Winit isn't the only Rust library trying it's hardest to be cross-platform.

My cleanup code records statistics for the user and some telemetry for us, it's not benign to users or us and definitely matters if it's reached or not.

> The academic nature of this concern is also what contributes to there being no right answer. It's easy to either strongly assert that you believe lots of developers will either fall afoul of this issue or assert the opposite.

I'm with you on that point, really not sure how to assert this one way or another.

---

> > Could call it `run_maybe_return()` It's a bit over-the-top but couldn't really be much clearer that the function might not return.
> 
> I would be perfectly happy with that.

Can we agree to that?